### DISCUSSION TITLE
API for traversing AST and help on System Adapters

### DISCUSSION THREAD

**Discussion by b4apple** [nokia/ntt#198] [Ideas]
Hi team,

Wondering if there is an API to traverse the ttcn AST as I want to write a interpreter for a VM and I think this would be a great start. Also, are there any system adapters available which I can plug into the VM or do I take a swig of coffee and do it on my own ?

Thanks,
Regards,
Aritro Shome

**Comment by 5nord** [comment]
**AST API**

We were experimenting with the syntax trees a long time and therefore have no API, but various flavors for syntax trees:
* a low-level abstract syntax tree in `go/ast`-style: https://pkg.go.dev/github.com/nokia/ntt@v0.14.3/ttcn3/ast
You traverse it manually or with `ast.Inspect`, `ast.Apply' or `ast.Walk`
* a `ttcn3.Tree` wrapper around the low-level AST, providing parents, caching, indexing, ...
* a serialized JSON tree for third party code generators used by `ttcn3c`

I am currently preparing a finalized API which will be released in a few weeks.


**System Adapters**

I don't have a proof of concept for adapters, yet. I am thinking about a RPC-based plugin architecture plus some tools for code generation. `ttcn3-bindgen` maybe?


Sorry, for the late answer.
